# Lab 1 - Introduction to cloud and local llm model

In [1]:
import ollama

In [2]:
response = ollama.chat(
    model='llama3.2:1b',
    messages=[
        {'role': 'user', 'content': 'What is an intelligent system?'}
    ]
)

In [3]:
print(response['message']['content'])

An intelligent system is a complex and dynamic entity that exhibits the ability to perceive, reason, learn, communicate, and interact with its environment in a way that demonstrates a level of cognitive sophistication. Intelligent systems are often characterized by their ability to process and analyze vast amounts of data, recognize patterns, and make decisions based on this analysis.

Intelligent systems can take many forms, including:

1. Artificial Intelligence (AI): This refers to the development of computer programs that can perform tasks that typically require human intelligence, such as learning, problem-solving, and decision-making.
2. Machine Learning: This is a subset of AI that involves the ability of a system to learn from experience and improve its performance over time without being explicitly programmed.
3. Human-Computer Interaction (HCI): This refers to the design and development of systems that interact with humans in a natural and intuitive way, often using visual in

In [4]:
from google import genai

In [5]:
client = genai.Client()  # reads GEMINI_API_KEY from my environment - the key is not stored in the notebook

In [6]:
response = client.models.generate_content(
    model="gemini-3.5-flash",
    contents="What is an intelligent system?"
)

In [7]:
print(response.text)

An **intelligent system** is a machine, software, or cyber-physical system that can perceive its environment, process information, learn from experiences, and take autonomous actions to achieve specific goals. 

Unlike traditional computer programs, which strictly follow pre-written, rigid instructions (if-then rules), an intelligent system can **adapt** to new situations, handle uncertainty, and improve its performance over time without direct human intervention.

---

### The Core Loop of an Intelligent System
Most intelligent systems operate on a continuous loop consisting of four main steps:

1. **Perceive (Sensing):** The system gathers data from its environment. This can be through physical sensors (cameras, microphones, GPS, thermometers) or digital data inputs (database feeds, user clicks).
2. **Reason (Processing):** The system analyzes the data. It uses algorithms, logic, and patterns to understand what the data means, predict future states, and evaluate potential actions.
3.

In [8]:
import pandas as pd
import time # to add time wrapper for latency call

In [9]:
df = pd.read_csv("dataset/student-mat.csv")
df = df.head(20) # keep it small for the lab

## Testing small dataset

In [10]:
# Using ollama (local)
def get_gender_count_ollama(gender):
    start = time.time()
    values = list(gender)
    response = ollama.chat(
        model='llama3.2:1b',
        messages=[
            {'role': 'user', 'content': f"Here is the 'sex' column from a dataset: {values}\n\nHow many Male and Female are in this dataset?"}
        ]
    )
    end = time.time()

    latency = end - start
    print(f"Latency: {latency:.2f}s")

    return response['message']['content'].strip()

In [11]:
gender_count_ollama = get_gender_count_ollama(df['sex'])

Latency: 1.32s


In [12]:
print(gender_count_ollama)

To determine the number of males and females in the dataset, you can count the occurrences of each sex.

There are 8 observations that are Female ('F') and 9 observations that are Male ('M'). 

Therefore, there are 8 Males and 9 Females in this dataset.


In [13]:
# Google (cloud)
def get_gender_count_gemini(gender):
    start = time.time()
    values = list(gender)
    response = client.models.generate_content(
        model="gemini-3.5-flash",
        contents=f"Here is the 'sex' column from a dataset: {values}\n\nHow many Male and Female are in this dataset?"
    )
    end = time.time()
    latency = end - start
    print(f"Latency: {latency:.2f}s")

    return response.text.strip()

In [14]:
gender_count_gemini = get_gender_count_gemini(df['sex'])

Latency: 3.18s


In [15]:
print(gender_count_gemini)

In this dataset, there are:

* **Female (F):** 11
* **Male (M):** 9

**Total:** 20 entries


## Testing model Reasoning

In [16]:
# Ollama (Reasoning)
def get_apples_reason_ollama():
    start = time.time()
    response = ollama.chat(
        model='llama3.2:1b',
        messages=[
            {'role': 'user', 'content': "Think step by step, then give your final answer."
            "\n\nQuestion: A store has 3 boxes of 12 apples each. 5 apples are rotten. "
            "How many apples remain?"}
        ]
    )
    end = time.time()
    latency = end - start
    print(f"Latency: {latency:.2f}s")

    return response['message']['content'].strip()

In [17]:
apples_ollama = get_apples_reason_ollama()

Latency: 1.07s


In [18]:
print(apples_ollama)

To find the total number of apples, I will multiply the number of boxes by the number of apples in each box:

3 boxes * 12 apples/box = 36 apples

Then, to find out how many apples remain after 5 are rotten, I will subtract the number of rotten apples from the total:

36 apples - 5 apples = 31 apples

The answer is 31.


In [19]:
# Ollama (Non-Reasoning)
def get_apples_answer_ollama():
    start = time.time()
    response = ollama.chat(
        model='llama3.2:1b',
        messages=[
            {'role': 'user', 'content': "Question: A store has 3 boxes of 12 apples each. 5 apples are rotten. "
            "How many apples remain?"}
        ]
    )
    end = time.time()
    latency = end - start
    print(f"Latency: {latency:.2f}s")

    return response['message']['content'].strip()

In [20]:
apples_answer_ollama = get_apples_answer_ollama()

Latency: 0.91s


In [21]:
print(apples_answer_ollama)

To find the total number of apples, multiply the number of boxes by the number of apples in each box.

3 boxes * 12 apples/box = 36 apples

Then subtract the number of rotten apples:

36 apples - 5 rotten apples = 31 apples

So, there are 31 apples remaining.


In [22]:
# Gemini (Reasoning)
def get_apples_reason_gemini():
    start = time.time()
    response = client.models.generate_content(
        model="gemini-3.5-flash",
        contents="Think step by step, then give your final answer."
            "\n\nQuestion: A store has 3 boxes of 12 apples each. 5 apples are rotten. "
            "How many apples remain?"
    )
    end = time.time()
    latency = end - start
    print(f"Latency: {latency:.2f}")

    return response.text.strip()

In [23]:
apples_gemini = get_apples_reason_gemini()

Latency: 2.09


In [24]:
print(apples_gemini)

To find out how many apples remain, we can solve this step-by-step:

**Step 1: Find the total number of apples.**
The store has 3 boxes, and each box contains 12 apples. 
$$\text{Total apples} = 3 \times 12 = 36\text{ apples}$$

**Step 2: Subtract the rotten apples.**
There are 5 rotten apples that need to be removed.
$$\text{Remaining apples} = 36 - 5 = 31\text{ apples}$$

**Final Answer:**
There are **31** apples remaining.


In [25]:
# Gemini (Non-Reasoning)
def get_apples_answer_gemini():
    start = time.time()
    response = client.models.generate_content(
        model="gemini-3.5-flash",
        contents="Question: A store has 3 boxes of 12 apples each. 5 apples are rotten. "
            "How many apples remain?"
    )
    end = time.time()
    latency = end - start
    print(f"Latency: {latency:.2f}")

    return response.text.strip()

In [26]:
apples_answer_gemini = get_apples_answer_gemini()

Latency: 1.67


In [27]:
print(apples_answer_gemini)

To find out how many apples remain, you can follow these steps:

1. **Calculate the total number of apples:** 
   3 boxes × 12 apples per box = 36 apples

2. **Subtract the rotten apples:** 
   36 total apples - 5 rotten apples = 31 apples

There are **31** apples remaining.


## Using GradioUI for interactive chat

In [28]:
import gradio as gr

c:\Users\bryan\anaconda3\envs\ollama_venv\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [29]:
def chat_with_ollama(message, history):
    response = ollama.chat(
        model='llama3.2:1b',
        messages=[
            {'role': 'user', 'content': message}
        ]
    )
    return response['message']['content']

In [ ]:
demo = gr.ChatInterface(
    fn=chat_with_ollama,
    title="Week 1: Local LLM Chat (Ollama)",
    description="Chatting with llama3.2:1b running via Ollama"
)

demo.launch()

# Checkpoint Questions
## What's the practical difference in setup complexity between local and cloud?
The difference is that cloud would require you to go to the official website and download it. Then it is ready to use.

For local, you would need to install it in a IDE using the command prompt. Then you would need a few lines of code to test if it works.

|Local|Cloud|
|-----|-----|
|Hosted on your local computer|Hosted on a company super computers|
|Uses personal computer processor and memory|Accessible via the internet|
|Has smaller parameters to fit onto the computer chip|Has many parameters, giving it maximum knowledge|
|Provides data privacy and no subscription cost|Higher speed, accuracy and solving most complex problems|
|Output quality is poor and slower as it is using your CPU|Data must be sent over the internet, raising privacy and security concern|
|Handle highly sensitive tasks like drafting confidential legal documents or private note taking|Handle deep market research, advanced coding or generated content|

## Where did the two models disagree, and why might that happen?
During the counting task for the different gender, the Ollama model had the total count of 17 and the Gemini model had the total count of 20. The Gemini model had the correct answer. This could be concluded that the Ollama model (local) has lower parameters to handle math problems.

For the reasoning task, both models explained their reasoning process, but the Gemini model also guides the user how to solve the problem manually.

||Ollama|Gemini|
|-|-----|------|
|Latency|0.91 - 1.07|1.66 - 2.07|
|Cost|Free|Free/Pay|
|Output Quality|Basic response.|Strong reasoning and straight to the point|